tensorpipe/test/channel/channel_test.h

```diff
@@ -203,39 +203,51 @@ class ChannelTestCase {
 
 template <typename TBuffer>
 class ClientServerChannelTestCase : public ChannelTestCase<TBuffer> {
-  std::vector<std::future<std::shared_ptr<tensorpipe::transport::Connection>>>
-  accept(
-      std::shared_ptr<tensorpipe::transport::Listener> listener,
-      size_t numConnections) {
-    auto promises = std::make_shared<std::vector<
-        std::promise<std::shared_ptr<tensorpipe::transport::Connection>>>>(
+  std::future<std::pair<
+      tensorpipe::Error,
+      std::vector<std::shared_ptr<tensorpipe::transport::Connection>>>>
+  accept(tensorpipe::transport::Listener& listener, size_t numConnections) {
+    auto promise = std::make_shared<std::promise<std::pair<
+        tensorpipe::Error,
+        std::vector<std::shared_ptr<tensorpipe::transport::Connection>>>>>();
+    auto connections = std::make_shared<
+        std::vector<std::shared_ptr<tensorpipe::transport::Connection>>>(
         numConnections);
+    auto connectionsMutex = std::make_shared<std::mutex>();
     for (size_t i = 0; i < numConnections; ++i) {
-      listener->accept(
-          [&, promises](
-              const tensorpipe::Error& error,
-              std::shared_ptr<tensorpipe::transport::Connection> connection) {
-            ASSERT_FALSE(error) << error.what();
-            connection->read([promises, connection](
-                                 const tensorpipe::Error& error,
-                                 const void* connIdBuf,
-                                 size_t length) {
-              ASSERT_FALSE(error) << error.what();
-              ASSERT_EQ(sizeof(uint64_t), length);
-              promises->at(*static_cast<const uint64_t*>(connIdBuf))
-                  .set_value(std::move(connection));
-            });
-          });
-    }
-
-    std::vector<std::future<std::shared_ptr<tensorpipe::transport::Connection>>>
-        futures;
-
-    for (auto& p : *promises) {
-      futures.push_back(p.get_future());
+      listener.accept([promise, connections, connectionsMutex](
+                          const tensorpipe::Error& error,
+                          std::shared_ptr<tensorpipe::transport::Connection>
+                              connection) mutable {
+        if (error) {
+          promise->set_value(std::make_pair(error, *connections));
```

Comment:
This risks calling the `set_value` method multiple times, which is an error and would throw. It could happen when multiple callbacks fire with error, which is likely, since once one of them does the listener will fire all the other ones too. Instead, we should probably add a flag (guarded by the mutex) telling us whether we've already set the future.